clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -509,7 +542,9 @@ public boolean isChannelReady(String id) {
     private KafkaChannel channelOrFail(String id) {
         KafkaChannel channel = this.channels.get(id);
         if (channel == null)
-            throw new IllegalStateException("Attempt to retrieve channel for which there is no open connection. Connection id " + id + " existing connections " + channels.keySet());
+            channel = this.closedChannels.get(id);
+        if (channel == null)
```

Comment:
Send failures on `closingChannels` are handled as ungraceful close, just as with normal channels, so the code is safe. But I agree it could be made more obvious. I have changed the method to take a flag and added a separate code path in `send` to make handling of closing channels more obvious.